### PR TITLE
admission: split plugins into validating-only and mutating+validating

### PR DESCRIPTION
**Pull request by sttts**
Prerequisite to enable admission webhooks by default.

### CODE REVIEW

File: pkg/project/apiserver/admission/nodeenv/admission.go [hunk 1]
```diff
@@ -67,12 +69,22 @@ func (p *podNodeEnvironment) Admit(a admission.Attributes) (err error) {
 		return apierrors.NewForbidden(resource, name, fmt.Errorf("pod node label selector conflicts with its project node label selector"))
 	}
 
-	// modify pod node selector = project node selector + current pod node selector
-	pod.Spec.NodeSelector = labelselector.Merge(projectNodeSelector, pod.Spec.NodeSelector)
+	if mutationAllowed {
+		// modify pod node selector = project node selector + current pod node selector
+		pod.Spec.NodeSelector = labelselector.Merge(projectNodeSelector, pod.Spec.NodeSelector)
+	}
```
Comment (liggitt) [inline pkg/project/apiserver/admission/nodeenv/admission.go]: needs an else to fail if the merged selector is not identical to the pod selector, right?

File: pkg/project/apiserver/admission/nodeenv/admission.go [hunk 2]
```diff
@@ -67,12 +69,22 @@ func (p *podNodeEnvironment) Admit(a admission.Attributes) (err error) {
 		return apierrors.NewForbidden(resource, name, fmt.Errorf("pod node label selector conflicts with its project node label selector"))
 	}
 
-	// modify pod node selector = project node selector + current pod node selector
-	pod.Spec.NodeSelector = labelselector.Merge(projectNodeSelector, pod.Spec.NodeSelector)
+	if mutationAllowed {
+		// modify pod node selector = project node selector + current pod node selector
+		pod.Spec.NodeSelector = labelselector.Merge(projectNodeSelector, pod.Spec.NodeSelector)
```
Comment (sttts) [inline pkg/project/apiserver/admission/nodeenv/admission.go]: should we check that the project node selector is contained?

Comment (sttts) [reply pkg/project/apiserver/admission/nodeenv/admission.go]: on validation I mean.

File: pkg/project/apiserver/admission/nodeenv/admission.go (hunk 1 above)
Comment (sttts) [reply pkg/project/apiserver/admission/nodeenv/admission.go]: hah, you were 1 min faster :)

Comment (sttts) [reply pkg/project/apiserver/admission/nodeenv/admission.go]: done

File: pkg/build/apiserver/admission/jenkinsbootstrapper/admission.go [hunk 3]
```diff
@@ -58,7 +59,7 @@ func NewJenkinsBootstrapper() admission.Interface {
 	}
 }
 
-func (a *jenkinsBootstrapper) Admit(attributes admission.Attributes) error {
+func (a *jenkinsBootstrapper) Validate(attributes admission.Attributes) error {
```
Comment (liggitt) [inline pkg/build/apiserver/admission/jenkinsbootstrapper/admission.go]: This one will be fun for dry run

File: pkg/security/apiserver/admission/sccadmission/scc_exec.go [hunk 4]
```diff
@@ -25,21 +25,21 @@ func RegisterSCCExecRestrictions(plugins *admission.Plugins) {
...
-// sccExecRestrictions is an implementation of admission.Interface which says no to a pod/exec on
+// sccExecRestrictions is an implementation of admission.ValidationInterface which says no to a pod/exec on
 // a pod that the user would not be allowed to create
 type sccExecRestrictions struct {
 	*admission.Handler
 	constraintAdmission *constraint
 	client              kclientset.Interface
 }
 
-func (d *sccExecRestrictions) Admit(a admission.Attributes) (err error) {
```
Comment (liggitt) [inline pkg/security/apiserver/admission/sccadmission/scc_exec.go]: Double check validating admission gets called on exec requests… it's not a normal storage impl

Comment (liggitt) [reply pkg/security/apiserver/admission/sccadmission/scc_exec.go]: Do we have an e2e test that checks if this is active and effective?

**Comment by liggitt** [comment]
/lgtm
/hold
for question on sccexec

**Comment by liggitt** [comment]
also looks like relevant test failures

File: pkg/security/apiserver/admission/sccadmission/scc_exec.go (hunk 4 above)
Comment (sttts) [reply pkg/security/apiserver/admission/sccadmission/scc_exec.go]: `Validate` is called: https://github.com/kubernetes/kubernetes/blob/d1636b8019fa042eb1135263b00293b1c806b1d7/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go#L132

I added an integration test for SCCExecRestrictions.

**Comment by sttts** [comment]
Have extended the PR quite a lot by adding `Validate` funcs where we only had `Admit`. Also added tests for both `Admit` and `Validate` where it made sense.

**Comment by sttts** [comment]
@simo5 @enj can you take a look whether the changes are sane?

File: pkg/build/apiserver/admission/secretinjector/admission.go [hunk 5]
```diff
@@ -93,7 +104,13 @@ func (si *secretInjector) Admit(attr admission.Attributes) (err error) {
...
 		secretName := match.Cookie.(string)
 		glog.V(4).Infof(`secretinjector: matched secret "%s/%s" to buildconfig "%s"`, namespace, secretName, bc.GetName())
-		bc.Spec.Source.SourceSecret = &api.LocalObjectReference{Name: secretName}
+		if bc.Spec.Source.SourceSecret == nil || bc.Spec.Source.SourceSecret.Name != secretName {
+			if mutationAllowed {
+				bc.Spec.Source.SourceSecret = &api.LocalObjectReference{Name: secretName}
+			} else {
+				return admission.NewForbidden(attr, fmt.Errorf("mutated spec.source.sourceSecret, expected: %v, got %v", api.LocalObjectReference{Name: secretName}, bc.Spec.Source.SourceSecret))
+			}
+		}
```
Comment (simo5) [inline pkg/build/apiserver/admission/secretinjector/admission.go]: Isn't this changing behavior by suddenly forbidding stuff during Validation that was previously admitted ?

Comment (sttts) [reply pkg/build/apiserver/admission/secretinjector/admission.go]: It shouldn't be as long as nobody did dirty post-secretInjector mutation in admission webhooks, which we have disabled by default.

**Comment by sttts** [comment]
/hold cancel

@liggitt's question is resolved.

File: pkg/build/apiserver/admission/secretinjector/admission.go [hunk 6]
```diff
@@ -93,7 +104,13 @@ func (si *secretInjector) Admit(attr admission.Attributes) (err error) {
 	if match := urlpattern.Match(patterns, url); match != nil {
 		secretName := match.Cookie.(string)
 		glog.V(4).Infof(`secretinjector: matched secret "%s/%s" to buildconfig "%s"`, namespace, secretName, bc.GetName())
-		bc.Spec.Source.SourceSecret = &api.LocalObjectReference{Name: secretName}
+		if bc.Spec.Source.SourceSecret == nil || bc.Spec.Source.SourceSecret.Name != secretName {
```
Comment (deads2k) [inline pkg/build/apiserver/admission/secretinjector/admission.go]: Pretty sure this method returns early if `bc.Spec.Source.SourceSecret != nil`, so the first predicate is extraneous

Comment (deads2k) [reply pkg/build/apiserver/admission/secretinjector/admission.go]: and if it's always nil, then the second predicate will always npe.

I'm not seeing a job here for validation admission in this plugin.

Comment (sttts) [reply pkg/build/apiserver/admission/secretinjector/admission.go]: The func returns early on non-nil.

But on nil we still have to check whether `urlpattern.Match(patterns, url)` matches and then return the error during validation (a webhook could have reset the secret to nil).

The more fundamental question is whether we want to re-check at all in injection admission plugins. They are not security critical.

File: pkg/project/apiserver/admission/nodeenv/admission.go (hunk 2 above)
Comment (sttts) [reply pkg/project/apiserver/admission/nodeenv/admission.go]: yes, old question.

File: pkg/quota/apiserver/admission/runonceduration/admission.go [hunk 7]
```diff
@@ -90,17 +100,36 @@ func (a *runOnceDuration) Admit(attributes admission.Attributes) error {
 		return nil
 	}
 
-	appliedProjectLimit, err := a.applyProjectAnnotationLimit(attributes.GetNamespace(), pod)
+	newSpec := pod.Spec
```
[the remark this answers is not in the document]
Comment (sttts) [reply pkg/quota/apiserver/admission/runonceduration/admission.go]: done

**Comment by sttts** [comment]
@deads2k addressed comments.

**Comment by sttts** [comment]
> Can you comment on behavioral changes, what is expected, what is not expected ?

@simo5  we split the admission plugins into a mutating and a validating phase (we introduced that split in https://github.com/kubernetes/kubernetes/pull/54689 upstream, this PR does the corresponding work for all OpenShift admission plugins). This change allows us to call out to user defined admission webhook (compare https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#what-are-admission-webhooks), let them mutate objects and then at the end our admission plugins' validation is called. So we can be sure that admission is effective even though users can customize it.

So in short: behaviour should be changed of any of the plugins. Instead the calls to `Admit`+webhook+`Validate` should do the same as am `Admit` before. If the webhook mutates what `Admit` did, the `Validation` call should catch that and refuse.

**Comment by sttts** [comment]
Rebased.

**Comment by sttts** [comment]
@simo5 lgty?

**Comment by sttts** [comment]
@deads2k with jordan removed from OWNERS this lost the approval.

**Comment by sttts** [comment]
Looks like the gcp test failure is real: `StdErr: "Error from server (Forbidden): pods \"multicast-0\" is forbidden: exec operation is not allowed because the pod's security context exceeds your permissions: pods \"multicast-0\" is forbidden: unable to validate against any pod security policy: []"`

**Comment by sttts** [comment]
@simo5 @deads2k please check the SCCExecRestrictions commit. `SCC.Validate` is very strict in the sense that the SA/user being checked for exec would have access to the SCC that was chosen when creating the pod. `SCC.Admit` in contrast allow a switch of the SCC (which is stored in the annotation).

**Comment by sttts** [comment]
@deads2k @simo5 lgty?
